resteasy-core/src/main/java/org/jboss/resteasy/core/InjectorFactoryImpl.java

```diff
@@ -248,4 +265,73 @@ else if (useDefault)
          return null;
       }
    }
+
+    enum OptionalInjections {
+        OPT     (Optional.class, OptionalInjections::getTypeArgument, Optional::empty, Optional::of),
+        OINT    (OptionalInt.class, x -> Integer.class, OptionalInt::empty, v -> OptionalInt.of((Integer) v)),
+        OLONG   (OptionalLong.class, x -> Long.class, OptionalLong::empty, v -> OptionalLong.of((Long) v)),
+        ODOUBLE (OptionalDouble.class, x -> Double.class, OptionalDouble::empty, v -> OptionalDouble.of((Double) v)),
+        ;
+
+        static Map<Class<?>, OptionalInjections> optionalTypes;
+
+        static {
+            optionalTypes = Arrays.stream(OptionalInjections.values())
+                    .collect(Collectors.toMap(o -> o.optional, Function.identity()));
+        }
+
+        final Class<?> optional;
+        final Function<Type, Type> valueType;
+        final Supplier<Object> empty;
+        final Function<Object, Object> present;
+
+        OptionalInjections(final Class<?> optional, final Function<Type, Type> valueType,
+                           final Supplier<Object> empty, final Function<Object, Object> present)
+        {
+            this.optional = optional;
+            this.valueType = valueType;
+            this.empty = empty;
+            this.present = present;
+        }
+
+
+        static ValueInjector wrap(Type paramType, Function<Type, ValueInjector> injectorFactory) {
+            return Optional.ofNullable(optionalTypes.get(Types.getRawType(paramType)))
+                    .<ValueInjector>map(oi -> new DelegatingInjector(oi, injectorFactory.apply(oi.valueType.apply(paramType))))
+                    .orElseGet(() -> injectorFactory.apply(paramType));
+        }

```

Comment:
Hi @stevenschlansker,

Mostly this looks pretty cool. I'm still ruminating about wrap(). You're creating an Optional<T> from T and then extracting T, right? Is that the best way to do it? Maybe it will be clear in the morning. ;-)

-Ron

Reply:
Originally, a ValueInjector is responsible for creating the object that will be injected as the parameter.
This takes a type and a ValueInjector factory (loosely, the "original behavior").  If the injected type is an Optional type (as encoded in the enum "data table") this will "wrap" the original type in a DelegatingInjector that is responsible for adding the Optional wrapper.

If you have any suggestions on how to improve the clarity I'm happy to make improvements!